keps/sig-apps/4368-support-managed-by-label-for-batch-jobs/README.md

```diff
@@ -0,0 +1,1158 @@
+
+<!--
+What other approaches did you consider, and why did you rule them out? These do
+not need to be as detailed as the proposal, but should include enough
+information to express the idea and why it was not acceptable.
+-->
+
+### Using label selectors
+
+We consider using label selectors by the Job controller to identify the subset
+of jobs it should watch. This could result in smaller memory usage.
+
+**Reasons for discarding/deferring**
+
+First, We use shared-informers (so that all core k8s controllers see all objects), then
+we cannot make the memory saving this way.
+
+Second, there is no "OR" logic in label selectors, however, the built-in Job
+controller needs to sync jobs in two cases:
+1. old jobs without the label
+2. new jobs with the label equal to `job-controller.k8s.io`
+
+This means we would need to go via a difficult process of ensuring all jobs
+have the label, or listen on events from two informers. In any case, the use of
+label-selectors is significantly more complicated than the skip `if` inside the
+`syncJob`, and does not allow for big memory gain.
+
+### Alternative ideas to improve debuggability
+
+#### Condition to indicate Job is skipped
+
+In order to inform the user that a job is skipped from synchronization we
+could add a dedicated condition, say `ManagedBy`, indicating that the job is
+skipped by the built-in controller.
+
+**Reasons for discarding/deferring**
+
+- Since the Job label is immutable, then the usability of the condition is limited,
+because the timestamp of the other fields will not bring extra debugging value.
+- Conceptually, we want to give full ownership of the Job object to the other
+job controller, objects mutated by two controllers could actually make debugging
+more involving.
+- The MultiKueue controller would have to non-trivially reconcile the Job Status.
+If it just blindly mirrored the status from the worker cluster (as currently
+planned), then it would remove the condition. Other controllers would need to be
+careful not to remove the condition either.
+- It requires extra request per job, and risks conflicts for the status Update
+requests.
+
+Additionally, notice that the analogous situation takes place when `spec.schedulerName`
+does not match a custom scheduling profile. There is no condition indicating that.
+
+#### Event indicating the Job is skipped
+
+Job controller could emit event on the Job creation event indicating the Job
+is synchronized by a custom controller. This would not run into the issue with
+controllers conflicting on status updates.
+
+**Reasons for discarding/deferring**
+
+Events have expiration time, which is potentially cloud-provider dependent.
+It makes them not that useful to debug situations when the Job didn't make
+progress for long time. So, they would not give a reliable signal for debugging
+based on playbooks.
+
+Renewing the even on every Job update seems excessive from the performance
+perspective.
+
```

Comment:
```suggestion

### Use `job.kubernetes.io/managed-by`

```

**But I'd swap this round**. `job.kubernetes.io/managed-by` should be the main proposal unless the
[generic label](https://github.com/kubernetes/enhancements/pull/4370/files#r1433055461) idea wins out. Using plain `managed-by` should be what we list as an alternative and then discard.

Why? Because “Kubernetes' official labels are always public” is much easier to teach than ”Kubernetes labels are public, except for X, Y and Z”. The more precedents we set, the more we encourage our project and third parties to use private labels with the eventual risk of key collision and heartache.

Reply:
Yes, actually the proposed label full name is `batch.kubernetes.io/managed-by` for there reasons. More details in the answer to https://github.com/kubernetes/enhancements/pull/4370#discussion_r1467397596.

Regarding "batch" vs. "job" prefix:

We prefer "batch" prefix for consistency with current labels used by the Job controller (see [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/batch/types.go#L29-L46)).

The "batch" prefix fits well for the use case of supporting this label
in other batch Job CRDs supported by Kueue, such as JobSet, MPIJob, RayJob, etc.

I put these considerations in the new Alternatives section. PTAL.